src/backend/columnar/columnar_tableam.c

```diff
@@ -529,12 +529,39 @@ columnar_index_fetch_tuple(struct IndexFetchTableData *sscan,
 	}
 
 	uint64 rowNumber = tid_to_row_number(*tid);
-	if (!ColumnarReadRowByRowNumber(scan->cs_readState, rowNumber, slot->tts_values,
-									slot->tts_isnull))
+	StripeMetadata *stripeMetadata =
+		FindStripeWithMatchingFirstRowNumber(columnarRelation, rowNumber, snapshot);
+	if (!stripeMetadata)
 	{
+		/* it is certain that tuple with rowNumber doesn't exist */
 		return false;
 	}
 
+	if (StripeIsFlushed(stripeMetadata) &&
+		!ColumnarReadRowByRowNumber(scan->cs_readState, rowNumber,
+									slot->tts_values, slot->tts_isnull))
+	{
+		/*
+		 * FindStripeWithMatchingFirstRowNumber doesn't verify upper row
+		 * number boundary of found stripe. For this reason, we didn't
+		 * certainly know if given row number belongs to one of the stripes.
+		 */
+		return false;
+	}
+
+	if (!StripeIsFlushed(stripeMetadata))
+	{
+		/*
```

Comment:
If the snapshot is not SnapshotDirty, it looks like that will be caught later, right? It might be good to Assert here just for documentation.

Reply:
> it looks like that will be caught later, right?

yes

it might be good to Assert here just for documentation.

👍